snappy-internal.h

```diff
@@ -36,13 +36,44 @@
 namespace snappy {
 namespace internal {
 
+#if !defined(SNAPPY_HAVE_BMI2)
+// __BMI2__ is defined by GCC and Clang. Visual Studio doesn't target BMI2
+// specifically, but it does define __AVX2__ when AVX2 support is available.
+// Fortunately, AVX2 was introduced in Haswell, just like BMI2.
+//
+// BMI2 is not defined as a subset of AVX2 (unlike SSSE3 and AVX above). So,
+// GCC and Clang can build code with AVX2 enabled but BMI2 disabled, in which
+// case issuing BMI2 instructions results in a compiler error.
+#if defined(__BMI2__) || (defined(_MSC_VER) && defined(__AVX2__))
+#define SNAPPY_HAVE_BMI2 1
+#else
+#define SNAPPY_HAVE_BMI2 0
+#endif
+#endif  // !defined(SNAPPY_HAVE_BMI2)
+
+#if SNAPPY_HAVE_BMI2
+// Please do not replace with <x86intrin.h>. or with headers that assume more
+// advanced SSE versions without checking with all the OWNERS.
+#include <immintrin.h>
+#endif
+
+#if (SNAPPY_HAVE_SSSE3 || SNAPPY_HAVE_NEON)
+#define SNAPPY_HAVE_VECTOR_BYTE_SHUFFLE 1
+#endif
 #if SNAPPY_HAVE_VECTOR_BYTE_SHUFFLE
 #if SNAPPY_HAVE_SSSE3
+#include <tmmintrin.h>
 using V128 = __m128i;
-#else
+#elif SNAPPY_HAVE_NEON
+#include <arm_neon.h>
 using V128 = uint8x16_t;
 #endif
 
+#if SNAPPY_HAVE_BMI2
+inline uint32_t zero_highbits_u32(uint32_t v, int n) {
```

Comment:
Could you please explain the benefit of wrapping `_bzhi_u32` with this function?

Until I learn more, I'm inclined to suggest that we remove the abstraction. It's currently used in exactly one place, and it's only defined when `SNAPPY_HAVE_BMI2`. So, it's effectively a less-known alias for the `_bzhi_u32` intrinsic.

Reply:
Well received, will update.